mm/2s2h/Enhancements/Items/AmmoBuyback.cpp

```diff
@@ -0,0 +1,384 @@
+#include <libultraship/bridge/consolevariablebridge.h>
+#include "2s2h/GameInteractor/GameInteractor.h"
+#include "2s2h/ShipInit.hpp"
+#include "2s2h/CustomMessage/CustomMessage.h"
+#include "2s2h/Enhancements/Enhancements.h"
+#include "message_data_fmt_nes.h"
+
+extern "C" {
+#include "overlays/actors/ovl_En_Fsn/z_en_fsn.h"
+
+void EnFsn_MakeOffer(EnFsn* thisx, PlayState* play);
+void EnFsn_GiveItem(EnFsn* thisx, PlayState* play);
+void EnFsn_StartBuying(EnFsn* thisx, PlayState* play);
+void EnFsn_ResumeInteraction(EnFsn* enFsn, PlayState* play);
+void Player_StartTalking(PlayState* play, Actor* actor);
+void Player_StopCutscene(Player* player);
+}
+
+#define CVAR_NAME "gEnhancements.Items.AmmoBuyback"
+#define CVAR CVarGetInteger(CVAR_NAME, AMMO_BUYBACK_VANILLA)
+
+#define TEXT_ID_FSN_REJECTION 0x29CF // "Sorry, but I can't sell that here."
+#define TEXT_ID_FSN_OFFER 0x29EF     // Price offer dialogue
+#define MAX_AMMO_SELL_QUANTITY 99
+#define STICK_THRESHOLD 30
+#define REPEAT_DELAY_INITIAL 10
+#define REPEAT_DELAY_CONTINUOUS 2
+
+static struct {
+    EnFsn* actor = nullptr;
+    ItemId itemId = ITEM_NONE;
+    s16 quantity = 0;
+    s16 price = 0;
+    s16 lastRupeesSelected = 0;
+    bool isInputActive = false;
+} sAmmoSale;
+
+static void Reset() {
+    sAmmoSale = {};
+}
+
+static s16 GetPricePerUnit(ItemId itemId) {
+    switch (itemId) {
+        case ITEM_BOW:
+            return 1;
+        case ITEM_BOMB:
+            return 3;
+        case ITEM_BOMBCHU:
+            return 4;
+        case ITEM_DEKU_STICK:
+            return 10;
+        case ITEM_DEKU_NUT:
+            return 3;
+        case ITEM_MAGIC_BEANS:
+            return 10;
+        case ITEM_POWDER_KEG:
+            return 50;
+        default:
+            return 0;
+    }
+}
+
+static const char* GetNameForDialogue(ItemId itemId) {
+    switch (itemId) {
+        case ITEM_BOW:
+            return "arrows";
+        case ITEM_BOMB:
+            return "bombs";
+        case ITEM_BOMBCHU:
+            return "bombchus";
+        case ITEM_DEKU_STICK:
+            return "deku sticks";
+        case ITEM_DEKU_NUT:
+            return "deku nuts";
+        case ITEM_MAGIC_BEANS:
+            return "magic beans";
+        case ITEM_POWDER_KEG:
+            return "powder kegs";
```

Comment:
Some of the capitalization may seem weird, but some of these items are always capitalized by style, and shop inventory is capitalized anyway.

```suggestion
        case ITEM_BOW:
            return "Arrows";
        case ITEM_BOMB:
            return "Bombs";
        case ITEM_BOMBCHU:
            return "Bombchus";
        case ITEM_DEKU_STICK:
            return "Deku Sticks";
        case ITEM_DEKU_NUT:
            return "Deku Nuts";
        case ITEM_MAGIC_BEANS:
            return "Magic Beans";
        case ITEM_POWDER_KEG:
            return "Powder Kegs";
```

Reply:
Yeh that's fine, but I might change them to red as well?

Reply:
That would make sense.